### PR TITLE
Iteration-level scheduler with geedy search and contrastive

### PR DESCRIPTION
**Pull request by KexinFeng**
This is outdated. See #803 

---
This PR is an iteration-level scheduler system, which is mainly implemented as a `class GreedySeqBatchScheduler()` and its parent class `SeqBatchScheduler`.  Another two important classes are: `SeqBatcher` and `Batch` (i.e. the `BatchList` in the java implementation)

See https://github.com/deepjavalibrary/djl/pull/2572 for detailed description of them.

To run this feature, see `test_scheduler.py`, which covers basically all the code. (Ignore the beam search and constrastive search part for now. They will be moved to a different PR)

Here are the main features:
1. Iteration-level adding request and request existing, with padding and pad trimming
2. Enable resusable prefix kv_cache input for init_forward. The kv_cache is pre-calculated.
3. Greedy search

Implementation directory
--- 
There are three major classes: `SeqBatchScheduler`, `SeqBatcher` and `Batch` and one major unit test: `test_greedy_scheduler` in this PR. See also https://github.com/deepjavalibrary/djl/pull/2572.

```python
"""
This is a scheduler, serving as an API to the consumer of the systme, allowing for three major actions: initForward, addBatch, fastForward, collectResults. 
"""
class SeqBatchScheduler(ABC):

    def __init__(self, lm_block: LMBlock, config: SearchConfig):
        self.lm_block = lm_block
        self.results = {}
        self.seq_batcher = None
        self.config = config
        self.search_configs = None

    @abstractmethod
    def init_forward(self, input_ids: torch.Tensor, batch_uids: torch.Tensor,
                     kv_cache: Union[Tuple, None]) -> SeqBatcher:

    def increment_forward(self, count: int) -> torch.Tensor:

    @abstractmethod
    def inference_call(self) -> torch.Tensor:
        pass

    def add_request(self, batch_uids: torch.Tensor,
                    input_ids: torch.Tensor,
                    search_configs: Dict[int, SearchConfig] = None,
                    kv_cache: Union[Tuple, None] = None):

    def collect_results(self):

```

```python
"""
This is a data class, which stores the search state (Batch), the control variables (eg seqLength, offSets, etc), and batch operations like merge_batch and trim()
"""
class SeqBatcher(object):

    def __init__(self, batch: Batch, request_uids: torch.Tensor,
                 offsets: torch.Tensor):
        self.batch = batch
        self.request_uids = request_uids
        self.offsets = offsets
        self.exit_index_end_position = {}

        past_key_values_size = batch.past_attention_mask.size()
        self.batch_size = past_key_values_size[0]
        self.seq_len = past_key_values_size[1]

    def add_batch(self, seq_batcher: SeqBatcher):
        return self.merge_symmetric(self, seq_batcher)

    def merge_symmetric(self, seq_batcher1: SeqBatcher, seq_batcher2: SeqBatcher):

    def collect_and_trim(self) -> Union[Dict[int, torch.Tensor], None]:

    def exit_criteria(self, output_ids: torch.Tensor, max_gen_seqlen: int,
                      eos_token_id: int):

    def seq_complete(self) -> bool:

```

```python
"""
Batch is a data class consisting of fields of tensors like past_output_ids, past_key_values,  past_attention_mask. It's a compact collection of variables that need to be updated in each incremental inference call.
"""
class Batch:

    def __init__(self,
                 past_output_ids: torch.Tensor = None,
                 past_attention_mask: torch.Tensor = None,
                 past_key_values=None,
                 logits: torch.Tensor = None):
        self.past_output_ids = past_output_ids
        self.past_key_values = past_key_values
        self.past_attention_mask = past_attention_mask
        self.logits = logits

    # merges another batch with itself.
    def merge(self, batch: Batch, seq_delta: int) -> Batch:

    def trim(self, keep_indices: torch.Tensor, trim_seq_len: int):

```

Miscellany
---
`assemble_prefix_kv_cache` and `nudge_to_squeeze_bubble_padding` are used only when a prefix kv_cache is provided. They are small augmented feature.

### CODE REVIEW

File: engines/python/setup/djl_python/scheduler/lm_block.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                       'output_hidden_states': config.get('output_hidden_states', True)}
+        model = GPT2LMHeadModel.from_pretrained(model_urls[0])  # it contains model.eval()
+        self.blocks = [model]
+
+    def forward(self, input: List[torch.tensor], past_key_values):
+        return self.blocks[0].forward(input_ids=input[0],
+                                      position_ids=input[1],
+                                      attention_mask=input[2],
+                                      past_key_values=past_key_values,
+                                      **self.config)
```
Comment (sindhuvahinis) [inline engines/python/setup/djl_python/scheduler/lm_block.py:75]: @KexinFeng  Could you raise this LM_BLOCK and its test case as a separate PR?

**Review by sindhuvahinis — APPROVED** [review]
